### PR TITLE
MAYA-122218 showing dactivated children

### PR DESCRIPTION
**Pull request by pierrebai-adsk**
Some code paths were ignoring deactivated children when they should not because hasChildren() was ignoring deactivated children.

### CODE REVIEW

**Comment by seando-adsk** [comment]
I'm not sure it is that simple. Did you look at the history of this file (and the pulledObjectHierarchy)? We (Krystian and I) have modified this code a few times to fix different bugs. Did you test that those scenarios (from prior bugs) are still fixed with these changes?

**Comment by pierrebai-adsk** [comment]
I looked at the most recent changes through git file history, like MAYA 113820.

File: lib/mayaUsd/ufe/UsdHierarchy.cpp [hunk 1]
```diff
@@ -124,12 +124,14 @@ bool UsdHierarchy::hasChildren() const
     //
     // I don't have data that proves we need to worry about performance in here,
     // so going after maintainability.
-    return !children().empty();
+    const bool isFilteringInactive = false;
+    return !createUFEChildList(getUSDFilteredChildren(fItem), isFilteringInactive).empty();
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdHierarchy.cpp:128]: If we are going with this change then we don't need the comment above anymore. Also have you looked at the changes Krystian made in the past? He was the the one who changed it to use children().

And have you looked at PulledObjectHierarchy and ProxyShapeHierarhcy - they both have similar code.

Comment (pierrebai-adsk) [reply lib/mayaUsd/ufe/UsdHierarchy.cpp:128]: I'll take a look at these classes, thanks.

Comment (pierrebai-adsk) [reply lib/mayaUsd/ufe/UsdHierarchy.cpp:128]: I re-read the comment and it needs to stay: what it cays is that calling create-children-list has extra costs since it will go through all children, but it is simpler to avoid duplicating the logic that filters the items out.

**Comment by pierrebai-adsk** [comment]
Only failure is the PF on OSX python 2 Maya 2020 failed to launch, but all others passed.

**Comment by pierrebai-adsk** [comment]
Th eonyl PF failure is the OSX python build that constantly fails to start.

File: test/lib/ufe/testContextOps.py [hunk 2]
```diff
@@ -362,8 +362,13 @@ def testAddNewPrimWithDelete(self):
         cmds.pickWalk(d='down')
         cmds.delete()
 
-        # The proxy shape should now have no UFE child items (since we skip inactive).
-        self.assertFalse(proxyShapehier.hasChildren())
+        # The proxy shape should now have no UFE child items (since we skip inactive)
+        # but hasChildren still reports true for inactive to allow the caller to then
+        # do conditional inactive filtering.
+        if mayaUtils.mayaMajorVersion() >= 2023:
```
Comment (seando-adsk) [inline test/lib/ufe/testContextOps.py:368]: I'm confused why this check is against a Maya version. What change is there in Maya that is affecting this?

**Review by seando-adsk — DISMISSED** [review]

File: lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp [hunk 3]
```diff
@@ -135,12 +135,17 @@ Ufe::SceneItem::Ptr ProxyShapeHierarchy::sceneItem() const { return fItem; }
...
+    if (!rootPrim.IsValid())
+        return false;
+
     // We have an extra logic in createUFEChildList to remap and filter
     // prims. Going this direction is more costly, but easier to maintain.
     //
     // I don't have data that proves we need to worry about performance in here,
     // so going after maintainability.
-    return !children().empty();
+    return !createUFEChildList(getUSDFilteredChildren(rootPrim), false /*filterInactive*/).empty();
```
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp:148]: I'm confused by this change.  Now hasChildren() is no longer a convenience for !children().empty(), and is creating a different filtered children list than what children() is doing.  Why is this correct?

File: test/lib/ufe/testContextOps.py (hunk 2 above)
Comment (ppt-adsk) [reply test/lib/ufe/testContextOps.py:368]: I'm confused as to why we changed the test at all.

Comment (pierrebai-adsk) [reply test/lib/ufe/testContextOps.py:368]: The behaviour of the delete command vs prim changed from deactibvating to really deleting. When deactivated, hasChildren() returns true.

File: lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp (hunk 3 above)
Comment (pierrebai-adsk) [reply lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp:148]: It is now a convenience for filteredChildren(0 with inactive prim. Unlike children(), which has an alternative function taking a filter that allows specifying to filter deactivated children or not, hasChildren() has a single version. So hasChildren has to choose to report deactivated children or not. The old choice was causing problems because code relied on it even when showing deactivated prims.

If we don't fix it here, then all Maya versions will still have bad outliner behaviour. Changing it fixes the behaviour in all Maya versions that use UFE v2 when they use the latest MayaUSD plugin.

File: lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp [hunk 4]
```diff
@@ -133,6 +133,8 @@ const UsdPrim& ProxyShapeHierarchy::getUsdRootPrim() const
 
 Ufe::SceneItem::Ptr ProxyShapeHierarchy::sceneItem() const { return fItem; }
 
+#if defined(UFE_V4_FEATURES_AVAILABLE) && (UFE_MINOR_VERSION > 4 || UFE_PATCH_LEVEL > 3)
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/ProxyShapeHierarchy.cpp:136]: This should just use a single test like this:
```suggestion
#if (UFE_PREVIEW_VERSION_NUM >= 4004)
```

File: lib/mayaUsd/ufe/PulledObjectHierarchy.h [hunk 5]
```diff
@@ -58,7 +58,10 @@ class MAYAUSD_CORE_PUBLIC PulledObjectHierarchy : public Ufe::Hierarchy
     Ufe::SceneItem::Ptr sceneItem() const override;
     bool                hasChildren() const override;
     Ufe::SceneItemList  children() const override;
-    Ufe::SceneItemList  filteredChildren(const ChildFilter&) const override;
+#if defined(UFE_V4_FEATURES_AVAILABLE) && (UFE_MINOR_VERSION > 4 || UFE_PATCH_LEVEL > 3)
+    bool hasFilteredChildren(const ChildFilter&) const override;
+#endif
+    UFE_V2(Ufe::SceneItemList filteredChildren(const ChildFilter&) const override;)
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/PulledObjectHierarchy.h:64]: Why did you add the UFE_V2 here? It is not needed as this file is only included in Ufe v3

File: test/lib/ufe/testContextOps.py (hunk 2 above)
Comment (seando-adsk) [reply test/lib/ufe/testContextOps.py:368]: Okay then instead of using then the Maya test should also surround lines 363 (which does the delete) and the comment you wrote here should be added to the test. Then for Maya 2023 we should not use delete, but instead deactivate the prim.

File: test/lib/ufe/testChildFilter.py [hunk 6]
```diff
@@ -92,7 +92,7 @@ def testFilteredChildren(self):
         ball3Prim = mayaUsd.ufe.ufePathToPrim(ball3PathStr)
         ball3Prim.SetActive(False)
 
-        # Props should now have 5 children and ball3 should not be one of them.
+        # Props children should have 5 children and ball3 should not be one of them.
```
Comment (seando-adsk) [inline test/lib/ufe/testChildFilter.py:95]: Why did you remove "now". The comment was correct. At line 84 we test that there are 6 children and now we are testing that there are 5.

**Review by seando-adsk — CHANGES_REQUESTED** [review]

File: test/lib/ufe/testChildFilter.py (hunk 6 above)
Comment (pierrebai-adsk) [reply test/lib/ufe/testChildFilter.py:95]: I took multiple passes to update the test and did not realized I dropped a word there.

File: lib/mayaUsd/ufe/PulledObjectHierarchy.cpp [hunk 7]
```diff
@@ -52,14 +52,23 @@ PulledObjectHierarchy::Ptr PulledObjectHierarchy::create(
...
+{
+    return _mayaHierarchy->hasFilteredChildren(childFilter);
+}
+#endif
+
 bool PulledObjectHierarchy::hasChildren() const { return _mayaHierarchy->hasChildren(); }
 
 Ufe::SceneItemList PulledObjectHierarchy::children() const { return _mayaHierarchy->children(); }
 
+#ifdef UFE_V2_FEATURES_AVAILABLE
```
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/PulledObjectHierarchy.cpp:66]: The entire file is UFE v3 (in CMakeLists.txt), so this ifdef is redundant.

File: test/lib/ufe/testContextOps.py [hunk 8]
```diff
@@ -358,17 +358,20 @@ def testAddNewPrimWithDelete(self):
         self.assertEqual(len(proxyShapehier.children()), 1)
 
         # Using UFE, delete this new prim (which doesn't actually delete it but
-        # instead makes it inactive).
+        # instead makes it inactive). In Maya 2023 and vreater, delete really
```
Comment (ppt-adsk) [inline test/lib/ufe/testContextOps.py:361]: "vreater" --> "greater"

File: test/lib/ufe/testContextOps.py [hunk 9]
```diff
@@ -358,17 +358,20 @@ def testAddNewPrimWithDelete(self):
...
-        cmds.delete()
+        if mayaUtils.mayaMajorVersion() >= 2023:
+            ufeItem = ufe.GlobalSelection.get().front()
+            item = usdUtils.getPrimFromSceneItem(ufeItem)
+            item.SetActive(False)
+        else:
+            cmds.delete()
 
         # The proxy shape should now have no UFE child items (since we skip inactive)
         # but hasChildren still reports true for inactive to allow the caller to then
```
Comment (ppt-adsk) [inline test/lib/ufe/testContextOps.py:372]: Comment and code don't match.

**Review by ppt-adsk — CHANGES_REQUESTED** [review]

File: test/testUtils/ufeUtils.py [hunk 10]
```diff
@@ -105,3 +105,26 @@ def ufeFeatureSetVersion():
     #   v0.2.20 (unreleased preview version 2).
     major = ufe.VersionInfo.getMajorVersion()
     return ufe.VersionInfo.getMinorVersion() if major == 0 else major
+
+def hasMinimumVersion(releaseVersions, unreleasedVersions):
```
Comment (ppt-adsk) [inline test/testUtils/ufeUtils.py:109]: You shouldn't need this.  See
https://github.com/Autodesk/maya-usd/pull/2096/files
on the approach we've chosen: we use the Maya preview release version and the UFE preview version to determine whether code should be compiled and tests run.  Then, when the next Maya preview release is released, we clean up ALL the preview release and UFE preview release conditions we're put in, and replace them with UFE feature set version checks.

This cleans up the code one preview release at a time so that we don't leave old unreleased version checks in the files.

**Review by ppt-adsk — CHANGES_REQUESTED** [review]

File: test/testUtils/ufeUtils.py (hunk 10 above)
Comment (pierrebai-adsk) [reply test/testUtils/ufeUtils.py:109]: In this case I till need to differentiate between 0.4.3 and 0.4.4 though, so I'll test that only intstead.

File: test/lib/ufe/testContextOps.py [hunk 11]
```diff
@@ -358,12 +359,26 @@ def testAddNewPrimWithDelete(self):
...
+
+        # The proxy shape should now have no UFE child items (since we skip inactive
+        # when returning the children list) but hasChildren still reports true in
+        # UFE version before 0.4.4 for inactive to allow the caller to do conditional
+        # inactive filtering, so we test that hasChildren is true for those versions.
+        ufeVersion = ufeUtils.fullVersion()
+        if (0, 4, 4) <= ufeVersion < (1, 0, 0):
+            self.assertFalse(proxyShapehier.hasChildren())
+        else:
+            self.assertTrue(proxyShapehier.hasChildren())
```
Comment (seando-adsk) [inline test/lib/ufe/testContextOps.py:381]: We have been doing this in a different way that allows us to find and remove these checks at a later date when a new Maya PR ships. Once the Maya PR ships with this new Ufe 0.4.x version we don't need a test on the specific Ufe version anymore - just a check against Ufe v4 is okay. So to find them we always use either the existing `ufeFeatureSetVersion` function or `UFE_PREVIEW_VERSION_NUM` and fix/remove them. So I think your test should be something like:
`if (os.getenv('UFE_PREVIEW_VERSION_NUM', '0000') > '4004'):`
This cmake variable is also set as an env var for all these Ufe tests.

**Comment by pierrebai-adsk** [comment]
The pm;y failure was on Windows 2020 python 2 where a schema test timed out for unknown reason. The test is entirely unrelated to my changes.

**Review by ppt-adsk — APPROVED** [review]

**Review by seando-adsk — APPROVED** [review]
